{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d0e18abf-1d86-4d0d-953f-7fca4e4510ed",
   "metadata": {},
   "source": [
    "# Step 1<br>\n",
    "### Input: path directories to each folder containing data<br>\n",
    "### Output: merged dataframes for analysis in Step 2<br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8d170886",
   "metadata": {},
   "outputs": [],
   "source": [
    "folders = [\n",
    "    '/Users/kateschnee/Library/CloudStorage/OneDrive-SharedLibraries-VUMC/Ellis, Sam - Collagen_Alignment_Project/Refined analysis pipeline/TR11_206',\n",
    "    '/Users/kateschnee/Library/CloudStorage/OneDrive-SharedLibraries-VUMC/Ellis, Sam - Collagen_Alignment_Project/Refined analysis pipeline/TR11_16184_2',\n",
    "    '/Users/kateschnee/Library/CloudStorage/OneDrive-SharedLibraries-VUMC/Ellis, Sam - Collagen_Alignment_Project/Refined analysis pipeline/TR11_21723',\n",
    "    '/Users/kateschnee/Library/CloudStorage/OneDrive-SharedLibraries-VUMC/Ellis, Sam - Collagen_Alignment_Project/Refined analysis pipeline/TR16_23542',\n",
    "    '/Users/kateschnee/Library/CloudStorage/OneDrive-SharedLibraries-VUMC/Ellis, Sam - Collagen_Alignment_Project/Refined analysis pipeline/TR11_18105'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "06fe5f77",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import csv\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import *\n",
    "from tabulate import tabulate\n",
    "import warnings\n",
    "import pandoc\n",
    "import time\n",
    "import math\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "31075b5b-3865-4ed9-a8ac-fe7857f47e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7a8da0a1-a6a2-4bec-a01b-41b2d015e81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_out = '/Users/kateschnee/Desktop/tmp/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "380566d6-4d99-47db-98aa-3c0053255241",
   "metadata": {},
   "outputs": [],
   "source": [
    "def progress_meter(i, n, e = 'blank'):\n",
    "    if e != 'blank':\n",
    "        print(f'{e}: {math.ceil(100 * i / n)}%', end = '\\r')\n",
    "    else:\n",
    "        print(f'{math.ceil(100 * i / n)}%', end = '\\r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1920302d-93bf-4f16-bc7e-66359a9748e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input: directory path for a sample (ex. /Users/.../TR11_206)\n",
    "# output: directory path for the same sample's ST data (ex. /Users/.../TR11_206/TR11_206_ST)\n",
    "def get_folder_data_path(path_in):\n",
    "    split_path = path_in.split(\"/\")\n",
    "    split_path.append(split_path[-1] + \"_ST\")\n",
    "    return \"/\".join(split_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "11f0f570-d0ff-4373-bfa8-9b910f89c94d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input: directory path for a sample's ST data (ex. /Users/.../TR11_206/TR11_206_ST)\n",
    "# reads through each of the files that ends in .tsv (stroma, epithelium, full, alignment)\n",
    "# we only care about stroma, epithelium, and alignment\n",
    "# loads the tsv data as a dataframe\n",
    "# output: updates the overarching dictionary all_data to include the dataframe\n",
    "# example: TR11_206: {stroma: stroma_dataframe}\n",
    "#                    {epithelium: epithelium_dataframe}\n",
    "#                    {alignment: alignment dataframe}\n",
    "#          TR11_16184_2: \n",
    "# ...\n",
    "\n",
    "def populate_first_dict(path_in):\n",
    "        \n",
    "    expt_name = path_in.split(\"/\")[-2]\n",
    "    dict_temp = {}\n",
    "    for filename in os.listdir(folder):\n",
    "    \n",
    "        # create temporary variables to store data\n",
    "        data_temp = []\n",
    "        cols_temp = []\n",
    "        \n",
    "        \n",
    "        # parse through the tsv files -- there should only be 4\n",
    "        if filename.endswith(\".tsv\"):\n",
    "            \n",
    "            key = filename.split('.')[0]\n",
    "                \n",
    "            df = pd.read_csv(folder + '/' + filename, sep = '\\t')\n",
    "\n",
    "            # make sure index loads, otherwise assign it\n",
    "            if df.index.tolist() == list(range(1, len(df.index) + 1)):\n",
    "                df.set_index('barcodes', inplace = True)\n",
    "            \n",
    "            if \"alignment\" in key:\n",
    "                dict_temp.update({\"alignment\": df})\n",
    "            \n",
    "            elif \"stroma\" in key:\n",
    "                dict_temp.update({\"stroma\": df})\n",
    "                                \n",
    "            elif \"epithelium\" in key:                \n",
    "                dict_temp.update({\"epithelium\": df})                \n",
    "\n",
    "    return dict_temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6ee0ab76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# all_data is a dictionary to store all information that the user inputs (see above functions)\n",
    "all_data = {}\n",
    "\n",
    "# iterate through each file directory that the user specifies, access its ST data, then load dataframe\n",
    "for item in folders:\n",
    "    \n",
    "    # get the path to the subfolder containing csv files\n",
    "    folder = get_folder_data_path(item)\n",
    "    \n",
    "    # get data into dictionary    \n",
    "    all_data.update({item.split('/')[-1]: populate_first_dict(folder)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "eb269a62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# access each sample's epithelial and stromal data\n",
    "# if a barcode is not in the alignment dataframe, we won't have x-y coordinates, so drop it\n",
    "for expt, spot in all_data.items():\n",
    "    for k, df in spot.items():\n",
    "        if k == 'epithelium' or k == 'stroma':\n",
    "            spot.update({k: df[df.index.isin((spot['alignment'].index.tolist()))]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "39f0c4c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input: stromal barcode (i), entire epithelial dataframe (epi_df), entire alignment dataframe (al_df)\n",
    "# accesses the stromal barcode's x-y position then compares it to every epithelial x-y\n",
    "# records lowest distance and retains that barcode\n",
    "def get_barcode_pair(base_barcode, comparison_df, alignment_df):\n",
    "    # arbitrary variables to get overwritten\n",
    "    min_dist = 9999999999\n",
    "    best_bc = ''\n",
    "    \n",
    "    # get base barcode's x-y data\n",
    "    base_x = alignment_df.loc[base_barcode, 'pxl_col_in_fullres']\n",
    "    base_y = alignment_df.loc[base_barcode, 'pxl_row_in_fullres']\n",
    "    \n",
    "    # iterate through every row in the comparison data\n",
    "    for comparison_barcode, comparison_data in comparison_df.iterrows():\n",
    " \n",
    "        # get comparison x-y\n",
    "        comparison_x = alignment_df.loc[comparison_barcode, 'pxl_col_in_fullres']\n",
    "        comparison_y = alignment_df.loc[comparison_barcode, 'pxl_row_in_fullres']\n",
    "        \n",
    "        # calculate distance\n",
    "        dist = math.sqrt((comparison_x - base_x) ** 2 + (comparison_y - base_y) ** 2)\n",
    "        \n",
    "        # potentially overwrite data if applicable\n",
    "        if dist < min_dist:\n",
    "            min_dist = dist\n",
    "            best_bc = comparison_barcode\n",
    "    \n",
    "    return min_dist, best_bc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ea73cf19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calls method to pair barcodes\n",
    "# inputs: base_region: the region acting as a \"base,\" either epithelium or stroma\n",
    "#         comparison_region: the region that is being iterated through every spot; the other region than above\n",
    "#         alignment_data: alignment values for the sample in question\n",
    "def parse_barcodes(base_region, comparison_region, alignment_data, expt, location):\n",
    "    i = 0\n",
    "    data_temp = []\n",
    "    \n",
    "    for index, row_data in base_region.iterrows():\n",
    "        i += 1\n",
    "        \n",
    "        # function to identify the epithelial partner\n",
    "        distance, barcode_match = get_barcode_pair(index, comparison_region, alignment_data)\n",
    "        \n",
    "        # record the stromal barcode (index), epithelial match (ebc), and the distance (d) for use in a df\n",
    "        data_temp.append([index, barcode_match, distance])\n",
    "        \n",
    "        # update progress for this sample\n",
    "        progress_meter(i, base_region.shape[0], expt + ' ' + location)\n",
    "    print()\n",
    "    df = pd.DataFrame(data_temp)\n",
    "    df.columns = [location + '_barcode', 'partner_barcode', 'distance']\n",
    "    df.set_index(location + '_barcode', inplace = True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4d449ce4-82e0-477a-a093-bf0563bd4399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temporary: shorten dataframes so faster\n",
    "# for expt, spot in all_data.items():\n",
    "#     for k, df in spot.items():\n",
    "#         if k == 'stroma':\n",
    "#             spot.update({k: df.iloc[0:10]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ec03c622",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TR11_206 stroma: 100%\n",
      "TR11_206 epithelium: 100%\n",
      "TR11_16184_2 stroma: 13%"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[31], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# for each sample, create a dataframe that acts as a legend\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m expt, spot \u001b[38;5;129;01min\u001b[39;00m all_data\u001b[38;5;241m.\u001b[39mitems():\n\u001b[0;32m----> 3\u001b[0m     stroma_pairings_df \u001b[38;5;241m=\u001b[39m parse_barcodes(spot[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstroma\u001b[39m\u001b[38;5;124m'\u001b[39m], spot[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mepithelium\u001b[39m\u001b[38;5;124m'\u001b[39m], spot[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124malignment\u001b[39m\u001b[38;5;124m'\u001b[39m], expt, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstroma\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      4\u001b[0m     epithelium_pairings_df \u001b[38;5;241m=\u001b[39m parse_barcodes(spot[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mepithelium\u001b[39m\u001b[38;5;124m'\u001b[39m] ,spot[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstroma\u001b[39m\u001b[38;5;124m'\u001b[39m], spot[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124malignment\u001b[39m\u001b[38;5;124m'\u001b[39m], expt, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mepithelium\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      5\u001b[0m     all_pairings_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([stroma_pairings_df, epithelium_pairings_df])\n",
      "Cell \u001b[0;32mIn[29], line 13\u001b[0m, in \u001b[0;36mparse_barcodes\u001b[0;34m(base_region, comparison_region, alignment_data, expt, location)\u001b[0m\n\u001b[1;32m     10\u001b[0m i \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;66;03m# function to identify the epithelial partner\u001b[39;00m\n\u001b[0;32m---> 13\u001b[0m distance, barcode_match \u001b[38;5;241m=\u001b[39m get_barcode_pair(index, comparison_region, alignment_data)\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# record the stromal barcode (index), epithelial match (ebc), and the distance (d) for use in a df\u001b[39;00m\n\u001b[1;32m     16\u001b[0m data_temp\u001b[38;5;241m.\u001b[39mappend([index, barcode_match, distance])\n",
      "Cell \u001b[0;32mIn[28], line 14\u001b[0m, in \u001b[0;36mget_barcode_pair\u001b[0;34m(base_barcode, comparison_df, alignment_df)\u001b[0m\n\u001b[1;32m     11\u001b[0m base_y \u001b[38;5;241m=\u001b[39m alignment_df\u001b[38;5;241m.\u001b[39mloc[base_barcode, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpxl_row_in_fullres\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m# iterate through every row in the comparison data\u001b[39;00m\n\u001b[0;32m---> 14\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m comparison_barcode, comparison_data \u001b[38;5;129;01min\u001b[39;00m comparison_df\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[1;32m     15\u001b[0m \n\u001b[1;32m     16\u001b[0m     \u001b[38;5;66;03m# get comparison x-y\u001b[39;00m\n\u001b[1;32m     17\u001b[0m     comparison_x \u001b[38;5;241m=\u001b[39m alignment_df\u001b[38;5;241m.\u001b[39mloc[comparison_barcode, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpxl_col_in_fullres\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     18\u001b[0m     comparison_y \u001b[38;5;241m=\u001b[39m alignment_df\u001b[38;5;241m.\u001b[39mloc[comparison_barcode, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpxl_row_in_fullres\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.13/site-packages/pandas/core/frame.py:1553\u001b[0m, in \u001b[0;36mDataFrame.iterrows\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1551\u001b[0m klass \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_constructor_sliced\n\u001b[1;32m   1552\u001b[0m using_cow \u001b[38;5;241m=\u001b[39m using_copy_on_write()\n\u001b[0;32m-> 1553\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m k, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalues):\n\u001b[1;32m   1554\u001b[0m     s \u001b[38;5;241m=\u001b[39m klass(v, index\u001b[38;5;241m=\u001b[39mcolumns, name\u001b[38;5;241m=\u001b[39mk)\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m   1555\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m using_cow \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr\u001b[38;5;241m.\u001b[39mis_single_block:\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.13/site-packages/pandas/core/frame.py:12664\u001b[0m, in \u001b[0;36mDataFrame.values\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m  12590\u001b[0m \u001b[38;5;129m@property\u001b[39m\n\u001b[1;32m  12591\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mvalues\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m np\u001b[38;5;241m.\u001b[39mndarray:\n\u001b[1;32m  12592\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m  12593\u001b[0m \u001b[38;5;124;03m    Return a Numpy representation of the DataFrame.\u001b[39;00m\n\u001b[1;32m  12594\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  12662\u001b[0m \u001b[38;5;124;03m           ['monkey', nan, None]], dtype=object)\u001b[39;00m\n\u001b[1;32m  12663\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m> 12664\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr\u001b[38;5;241m.\u001b[39mas_array()\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.13/site-packages/pandas/core/internals/managers.py:1694\u001b[0m, in \u001b[0;36mBlockManager.as_array\u001b[0;34m(self, dtype, copy, na_value)\u001b[0m\n\u001b[1;32m   1692\u001b[0m         arr\u001b[38;5;241m.\u001b[39mflags\u001b[38;5;241m.\u001b[39mwriteable \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m   1693\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1694\u001b[0m     arr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_interleave(dtype\u001b[38;5;241m=\u001b[39mdtype, na_value\u001b[38;5;241m=\u001b[39mna_value)\n\u001b[1;32m   1695\u001b[0m     \u001b[38;5;66;03m# The underlying data was copied within _interleave, so no need\u001b[39;00m\n\u001b[1;32m   1696\u001b[0m     \u001b[38;5;66;03m# to further copy if copy=True or setting na_value\u001b[39;00m\n\u001b[1;32m   1698\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m na_value \u001b[38;5;129;01mis\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mno_default:\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.13/site-packages/pandas/core/internals/managers.py:1727\u001b[0m, in \u001b[0;36mBlockManager._interleave\u001b[0;34m(self, dtype, na_value)\u001b[0m\n\u001b[1;32m   1724\u001b[0m \u001b[38;5;66;03m# error: Argument 1 to \"ensure_np_dtype\" has incompatible type\u001b[39;00m\n\u001b[1;32m   1725\u001b[0m \u001b[38;5;66;03m# \"Optional[dtype[Any]]\"; expected \"Union[dtype[Any], ExtensionDtype]\"\u001b[39;00m\n\u001b[1;32m   1726\u001b[0m dtype \u001b[38;5;241m=\u001b[39m ensure_np_dtype(dtype)  \u001b[38;5;66;03m# type: ignore[arg-type]\u001b[39;00m\n\u001b[0;32m-> 1727\u001b[0m result \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mempty(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshape, dtype\u001b[38;5;241m=\u001b[39mdtype)\n\u001b[1;32m   1729\u001b[0m itemmask \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mzeros(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m])\n\u001b[1;32m   1731\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m dtype \u001b[38;5;241m==\u001b[39m np\u001b[38;5;241m.\u001b[39mdtype(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mobject\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m na_value \u001b[38;5;129;01mis\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mno_default:\n\u001b[1;32m   1732\u001b[0m     \u001b[38;5;66;03m# much more performant than using to_numpy below\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# for each sample, create a dataframe that acts as a legend\n",
    "for expt, spot in all_data.items():\n",
    "    stroma_pairings_df = parse_barcodes(spot['stroma'], spot['epithelium'], spot['alignment'], expt, 'stroma')\n",
    "    epithelium_pairings_df = parse_barcodes(spot['epithelium'] ,spot['stroma'], spot['alignment'], expt, 'epithelium')\n",
    "    all_pairings_df = pd.concat([stroma_pairings_df, epithelium_pairings_df])\n",
    "    new_alignment = spot['alignment'].join(all_pairings_df, how = 'left')\n",
    "    new_alignment.dropna(inplace = True)\n",
    "    #print(f'Original alignment dataframe: {spot['alignment'].shape[0]}')\n",
    "    #print(f'Original epithelium dataframe: {spot['epithelium'].shape[0]}')\n",
    "    #print(f'Original stroma dataframe: {spot['stroma'].shape[0]}')\n",
    "    #print(f'Final alignment dataframe: {new_alignment.shape[0]}')\n",
    "    spot.update({'alignment': new_alignment})\n",
    "\n",
    "    for k, df in spot.items():\n",
    "        df.to_csv(path_out + expt + '_' + k + '.csv')\n",
    "    \n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e448e696",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # temporary: load the two data sets\n",
    "# all_data = {}\n",
    "# all_data.update({'TR11_206': {'epithelium': pd.read_csv(path_out + 'TR11_206_epithelium.csv', index_col = 0)}})\n",
    "# all_data['TR11_206']['stroma'] = pd.read_csv(path_out + 'TR11_206_stroma.csv', index_col = 0)\n",
    "# all_data['TR11_206']['alignment'] = pd.read_csv(path_out + 'TR11_206_alignment.csv', index_col = 0)\n",
    "# all_data.update({'TR11_16184_2': {'epithelium': pd.read_csv(path_out + 'TR11_16184_2_epithelium.csv', index_col = 0)}})\n",
    "# all_data['TR11_16184_2']['stroma'] = pd.read_csv(path_out + 'TR11_16184_2_stroma.csv', index_col = 0)\n",
    "# all_data['TR11_16184_2']['alignment'] = pd.read_csv(path_out + 'TR11_16184_2_alignment.csv', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac0dc185",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for expt, spot in all_data.items():\n",
    "#     for k, df in spot.items():\n",
    "\n",
    "#         if k != 'alignment':\n",
    "#             #df.columns.values[0] = 'barcode'\n",
    "#             df.drop('barcodes', axis = 1, inplace = True)\n",
    "#             spot.update({k: df})\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "806ce34b-da82-4c10-b940-fdaf75e11a9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_genes(dfe, dfs, gene_list):\n",
    "    gene_list.extend(dfe.columns)\n",
    "    gene_list.extend(dfs.columns)\n",
    "    return list(set(gene_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f2daf214-583b-4982-9dc2-ce2dc79debb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a list of every gene to account for differences between samples\n",
    "all_genes = []\n",
    "\n",
    "# parse through each sample's ST data that was loaded previously\n",
    "for expt, spot in all_data.items():\n",
    "\n",
    "    epi_df, stroma_df = spot['epithelium'], spot['stroma']\n",
    "    \n",
    "    # add any new genes to the growing list to account for differences across samples\n",
    "    all_genes = update_genes(epi_df, stroma_df, all_genes)\n",
    "\n",
    "    # create alignment data and account for the non-gene data\n",
    "    alignment_df = spot['alignment'].add_suffix('_non_gene')\n",
    "    alignment_df.rename(columns = {'alignment_non_gene': 'alignment'}, inplace = True)\n",
    "\n",
    "    # add genes that may be different with values of 0\n",
    "    for g in all_genes:\n",
    "        if g not in epi_df.columns:\n",
    "            epi_df[g] = list(np.zeros(epi_df.shape[0]))\n",
    "        if g not in stroma_df.columns:\n",
    "            stroma_df[g] = list(np.zeros(stroma_df.shape[0]))\n",
    "\n",
    "    spot.update({'alignment': alignment_df})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "eed42f5b-33a1-4aa2-a2da-cb93dee7c047",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input: either the epithelial or stromal dataframe (df1) and the alignment data (df2)\n",
    "def merge_df(df1, df2):\n",
    "    \n",
    "    # merge the dataframes horizontally and only include overlapping indices (barcodes)\n",
    "    new_df = pd.merge(df1, df2, left_index = True, right_index = True, how = 'inner')\n",
    "    \n",
    "    # remove the duplicate column \"barcode\" that might appear    \n",
    "    if 'barcodes' in new_df.columns:\n",
    "        new_df.drop('barcodes', axis = 1, inplace = True)\n",
    "    \n",
    "    \n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "c61500d5-6f25-4729-a1c3-deff48d8cd3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for expt, spot in all_data.items():\n",
    "    # merge the epithelium and stromal data with the alignment data\n",
    "    # now, we do not need to worry about the alignment dataframe\n",
    "\n",
    "    epi_df, stroma_df, alignment_df = spot['epithelium'], spot['stroma'], spot['alignment']\n",
    "    spot.update({'epithelium': merge_df(epi_df, alignment_df)})\n",
    "    spot.update({'stroma': merge_df(stroma_df, alignment_df)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "ee4a2ec6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TR11_206 stroma: (1964, 18035)\n",
      "TR11_206 epithelium: (852, 18035)\n",
      "TR11_206 alignment: (2816, 10)\n",
      "TR11_16184_2 stroma: (1467, 18049)\n",
      "TR11_16184_2 epithelium: (1581, 18049)\n",
      "TR11_16184_2 alignment: (3048, 10)\n",
      "TR11_21723 stroma: (2605, 18064)\n",
      "TR11_21723 alignment: (3750, 10)\n",
      "TR11_21723 epithelium: (1145, 18064)\n",
      "TR16_23542 stroma: (3333, 18064)\n",
      "TR16_23542 alignment: (4564, 10)\n",
      "TR16_23542 epithelium: (1231, 18064)\n",
      "TR11_18105 stroma: (2845, 18065)\n",
      "TR11_18105 alignment: (3276, 10)\n",
      "TR11_18105 epithelium: (431, 18065)\n"
     ]
    }
   ],
   "source": [
    "for expt, spot in all_data.items():\n",
    "    for k, df in spot.items():\n",
    "        print(f'{expt} {k}: {df.shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "bec25e87-f5f0-4ec8-aa71-fb95e294637b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_dataframe(expt, df_in, alignment_df, existing_df):\n",
    "    temp_df = df_in.reset_index()\n",
    "    temp_df.rename(columns = {'index': 'barcode'}, inplace = True)\n",
    "    temp_df.insert(0, 'experiment', [expt] * temp_df.shape[0])\n",
    "    if df.shape[0] == 0:\n",
    "        return temp_df\n",
    "    else:\n",
    "        return pd.concat([existing_df, temp_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "3787a445-0fe8-4b7b-84ef-c291d0f7aab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now, we want to combine the epithelial, stromal, and barcode-pairing dataframes across samples\n",
    "dfe = pd.DataFrame()\n",
    "dfs = pd.DataFrame()\n",
    "\n",
    "\n",
    "# iterate through each sample's data\n",
    "for expt, spot in all_data.items():\n",
    "\n",
    "    dfe = merge_dataframe(expt, spot['epithelium'], spot['alignment'], dfe)\n",
    "    dfs = merge_dataframe(expt, spot['stroma'], spot['alignment'], dfs)\n",
    "        \n",
    "dfe.set_index(['experiment', 'barcode'], inplace=True)\n",
    "dfs.set_index(['experiment', 'barcode'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "407d68fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfe.to_csv(path_out + 'dfe.csv', index = True)\n",
    "dfs.to_csv(path_out + 'dfs.csv', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "d4d9bb15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total runtime: 435 minutes and 16 seconds.\n"
     ]
    }
   ],
   "source": [
    "end_time = time.time()\n",
    "print(f'Total runtime: {math.floor((end_time - start_time)/60)} minutes and {math.floor((end_time - start_time)%60)} seconds.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "2050cfb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "partner_barcode_non_gene\n"
     ]
    }
   ],
   "source": [
    "for c in dfe.columns:\n",
    "    if 'barcode' in c:\n",
    "        print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "a29fcc02",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/m2/fxhwzr9d629bg5_5v13p70_00000gn/T/ipykernel_74342/2895656434.py:18: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  rho, pval = spearmanr(dfe[gene], dfs_aligned['alignment'])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         dfe_gene   dfs_gene       rho      pval\n",
      "0          SAMD11  alignment  0.012491  0.365996\n",
      "1           NOC2L  alignment -0.013606  0.324773\n",
      "2          KLHL17  alignment -0.011974  0.386174\n",
      "3         PLEKHN1  alignment -0.006535  0.636260\n",
      "4           PERM1  alignment  0.027966  0.042941\n",
      "...           ...        ...       ...       ...\n",
      "18051      NLGN4Y  alignment       NaN       NaN\n",
      "18052  AC007244.1  alignment       NaN       NaN\n",
      "18053       KDM5D  alignment       NaN       NaN\n",
      "18054      EIF1AY  alignment       NaN       NaN\n",
      "18055      TCEAL5  alignment       NaN       NaN\n",
      "\n",
      "[18056 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "dfs_aligned = dfs.loc[pd.MultiIndex.from_arrays([\n",
    "    dfe.index.get_level_values(\"experiment\"),\n",
    "    dfe[\"partner_barcode_non_gene\"]\n",
    "])].set_index(dfe.index)\n",
    "\n",
    "cols_to_avoid = []\n",
    "for c in dfe.columns:\n",
    "    if c.endswith('_non_gene'):\n",
    "        cols_to_avoid.append(c)\n",
    "    elif 'barcode' in c:\n",
    "        cols_to_avoid.append(c)\n",
    "cols_to_avoid.append('experiment')\n",
    "\n",
    "results = {}\n",
    "for gene in dfe.columns:\n",
    "    if gene not in cols_to_avoid:\n",
    "        try:\n",
    "            rho, pval = spearmanr(dfe[gene], dfs_aligned['alignment'])\n",
    "            results[(gene, 'alignment')] = (rho, pval)\n",
    "        except Exception as e:\n",
    "            results[(gene, 'alignment')] = ('NA', 'NA')\n",
    "\n",
    "# Tidy results table\n",
    "results_df = (\n",
    "    pd.DataFrame(results, index=[\"rho\", \"pval\"])\n",
    "    .T\n",
    "    .reset_index()\n",
    "    .rename(columns={\"level_0\": \"dfe_gene\", \"level_1\": \"dfs_gene\"})\n",
    ")\n",
    "\n",
    "\n",
    "results_df.to_csv(path_out + 'new_stroma_comparisons.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2716909a",
   "metadata": {},
   "outputs": [],
   "source": [
    "end_time = time.time()\n",
    "print(f'Total runtime: {math.floor((end_time - start_time)/60)} minutes and {math.floor((end_time - start_time)%60)} seconds.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
